### PR TITLE
reduce compiler warnings

### PR DESCRIPTION
**Pull request by asalmgren**
## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

File: Src/Particle/AMReX_NeighborList.H [hunk 1]
```diff
@@ -203,9 +203,9 @@ public:
 
         AMREX_FOR_1D ( np_real, i,
         {
-            int ix = (pstruct_ptr[i].pos(0)-plo[0])*dxi[0] - lo.x;
-            int iy = (pstruct_ptr[i].pos(1)-plo[1])*dxi[1] - lo.y;
-            int iz = (pstruct_ptr[i].pos(2)-plo[2])*dxi[2] - lo.z;
+            int ix = static_cast<int>((pstruct_ptr[i].pos(0)-plo[0])*dxi[0] - lo.x);
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int ix = static_cast<int>((pstruct_ptr[i].pos(0)-plo[0])*dxi[0]) - lo.x;
```

**Review by WeiqunZhang — APPROVED** [review]

**Comment by WeiqunZhang** [comment]
Do not merge!  My suggestion is not right.

File: Src/Particle/AMReX_NeighborList.H [hunk 2]
```diff
@@ -203,9 +203,9 @@ public:
 
         AMREX_FOR_1D ( np_real, i,
         {
-            int ix = (pstruct_ptr[i].pos(0)-plo[0])*dxi[0] - lo.x;
-            int iy = (pstruct_ptr[i].pos(1)-plo[1])*dxi[1] - lo.y;
-            int iz = (pstruct_ptr[i].pos(2)-plo[2])*dxi[2] - lo.z;
+            int ix = static_cast<int>((pstruct_ptr[i].pos(0)-plo[0])*dxi[0]) - lo.x;
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int ix = static_cast<int>(amrex::Math::floor((pstruct_ptr[i].pos(0)-plo[0])*dxi[0])) - lo.x;
```

File: Src/Particle/AMReX_NeighborList.H [hunk 3]
```diff
@@ -203,9 +203,9 @@ public:
...
+            int iy = static_cast<int>((pstruct_ptr[i].pos(1)-plo[1])*dxi[1]) - lo.y;
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int iy = static_cast<int>(amrex::Math::floor((pstruct_ptr[i].pos(1)-plo[1])*dxi[1])) - lo.y;
```

File: Src/Particle/AMReX_NeighborList.H [hunk 4]
```diff
@@ -203,9 +203,9 @@ public:
...
+            int iz = static_cast<int>((pstruct_ptr[i].pos(2)-plo[2])*dxi[2]) - lo.z;
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int iz = static_cast<int>(amrex::Math::floor((pstruct_ptr[i].pos(2)-plo[2])*dxi[2])) - lo.z;
```

File: Src/Particle/AMReX_NeighborList.H [hunk 5]
```diff
@@ -187,9 +187,9 @@ public:
                                       AMREX_ASSERT((p.pos(1)-plo[1])*dxi[1] - lo.y >= 0.0);,
                                       AMREX_ASSERT((p.pos(2)-plo[2])*dxi[2] - lo.z >= 0.0));
 
-                         return IntVect(AMREX_D_DECL(static_cast<int>((p.pos(0)-plo[0])*dxi[0] - lo.x),
-                                                     static_cast<int>((p.pos(1)-plo[1])*dxi[1] - lo.y),
-                                                     static_cast<int>((p.pos(2)-plo[2])*dxi[2] - lo.z)));
+                         return IntVect(AMREX_D_DECL(static_cast<int>((p.pos(0)-plo[0])*dxi[0]) - lo.x,
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
                         return IntVect(AMREX_D_DECL(static_cast<int>(amrex::Math::floor((p.pos(0)-plo[0])*dxi[0])) - lo.x,
```

File: Src/Particle/AMReX_NeighborList.H [hunk 6]
```diff
@@ -187,9 +187,9 @@ public:
...
+                                                     static_cast<int>((p.pos(1)-plo[1])*dxi[1]) - lo.y,
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
                                                     static_cast<int>(amrex::Math::floor((p.pos(1)-plo[1])*dxi[1])) - lo.y,
```

File: Src/Particle/AMReX_NeighborList.H [hunk 7]
```diff
@@ -187,9 +187,9 @@ public:
...
+                                                     static_cast<int>((p.pos(2)-plo[2])*dxi[2]) - lo.z));
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
                                                     static_cast<int>(amrex::Math::floor((p.pos(2)-plo[2])*dxi[2])) - lo.z));
```

File: Src/Particle/AMReX_NeighborList.H [hunk 8]
```diff
@@ -246,9 +246,9 @@ public:
 
         AMREX_FOR_1D ( np_real, i,
         {
-            int ix = (pstruct_ptr[i].pos(0)-plo[0])*dxi[0] - lo.x;
-            int iy = (pstruct_ptr[i].pos(1)-plo[1])*dxi[1] - lo.y;
-            int iz = (pstruct_ptr[i].pos(2)-plo[2])*dxi[2] - lo.z;
+            int ix = static_cast<int>((pstruct_ptr[i].pos(0)-plo[0])*dxi[0]) - lo.x;
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int ix = static_cast<int>(amrex::Math::floor((pstruct_ptr[i].pos(0)-plo[0])*dxi[0])) - lo.x;
```

File: Src/Particle/AMReX_NeighborList.H [hunk 9]
```diff
@@ -246,9 +246,9 @@ public:
...
+            int iy = static_cast<int>((pstruct_ptr[i].pos(1)-plo[1])*dxi[1]) - lo.y;
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int iy = static_cast<int>(amrex::Math::floor((pstruct_ptr[i].pos(1)-plo[1])*dxi[1])) - lo.y;
```

File: Src/Particle/AMReX_NeighborList.H [hunk 10]
```diff
@@ -246,9 +246,9 @@ public:
...
+            int iz = static_cast<int>((pstruct_ptr[i].pos(2)-plo[2])*dxi[2]) - lo.z;
```
Comment (atmyers) [inline Src/Particle/AMReX_NeighborList.H]:
```suggestion
            int iz = static_cast<int>(amrex::Math::floor((pstruct_ptr[i].pos(2)-plo[2])*dxi[2])) - lo.z;
```

**Review by atmyers — COMMENTED** [review]
Moving the location of the cast is subtly wrong. See the attached drawing. 

We actually did this once before: 

https://github.com/AMReX-Codes/amrex/pull/882
https://github.com/AMReX-Codes/amrex/pull/885

I think the cleanest thing is to move the casts but also add `amrex::Math:floor` calls, which is more like what we do when we bin particles in other parts of AMReX. The original code is equivalent to the version I suggest iff `(p.pos(1)-plo[1])*dxi[1] - lo.y >= 0.0`, which we assert for. 

You can apply the suggestions through the web interface if that's easier.
IMG_20210120_143103

**Pull request merged**